rasa/shared/core/domain.py

```diff
@@ -270,58 +275,124 @@ def merge(self, domain: Optional["Domain"], override: bool = False) -> "Domain":
+        ``actions``) are known to the Domain when the object is created.
+
+        List attributes like ``intents`` and ``actions`` are deduped
+        and merged. Single attributes are taken from `domain1` unless
+        override is `True`, in which case they are taken from `domain2`.
+        """
+        if not domain2:
+            return domain1
+
+        if not domain1:
+            return domain2
+
+        domain_dict = domain2
+        combined = domain1
 
         if override:
             config = domain_dict["config"]
             for key, val in config.items():
                 combined["config"][key] = val
 
-        if override or self.session_config == SessionConfig.default():
+        if override or domain2.get("session_config"):
             combined[SESSION_CONFIG_KEY] = domain_dict[SESSION_CONFIG_KEY]
 
-        duplicates: Dict[Text, List[Text]] = {}
-
-        # this code merges lists of dicts of intents
-        dict1 = {list(i.keys())[0]: i for i in combined[KEY_INTENTS]}
-        dict2 = {list(i.keys())[0]: i for i in domain_dict[KEY_INTENTS]}
-        duplicates[KEY_INTENTS] = extract_duplicates(dict1, dict2)
-        combined[KEY_INTENTS] = list(merge_dicts(dict1, dict2, override).values())
+        if combined.get(KEY_INTENTS) or domain_dict.get(KEY_INTENTS):
+            combined[KEY_INTENTS] = combined.get(KEY_INTENTS, [])
+            domain_dict[KEY_INTENTS] = domain_dict.get(KEY_INTENTS, [])
+            combined[KEY_INTENTS] = self.merge_lists_of_dicts(
+                combined[KEY_INTENTS], domain_dict[KEY_INTENTS], override
+            )
 
         # remove existing forms from new actions
-        for form in combined[KEY_FORMS]:
-            if form in domain_dict[KEY_ACTIONS]:
+        for form in combined.get(KEY_FORMS, []):
+            if form in domain_dict.get(KEY_ACTIONS, []):
                 domain_dict[KEY_ACTIONS].remove(form)
 
         for key in [KEY_ENTITIES, KEY_ACTIONS, KEY_E2E_ACTIONS]:
-            combined[key] = merge_lists(combined[key], domain_dict[key])
+            combined[key] = self.merge_lists(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
 
         for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
-            duplicates[key] = extract_duplicates(combined[key], domain_dict[key])
-            combined[key] = merge_dicts(combined[key], domain_dict[key], override)
+            combined[key] = self.merge_dicts(
+                combined.get(key, {}), domain_dict.get(key, {}), override
+            )
 
-        return self.__class__.from_dict(combined, duplicates)
+        return combined
+
+    @staticmethod
+    def merge_dicts(
+        tempDict1: Dict[Text, Any],
+        tempDict2: Dict[Text, Any],
+        override_existing_values: bool = False,
+    ) -> Dict[Text, Any]:
+        """Merges two dicts."""
+        if override_existing_values:
+            merged_dicts, b = tempDict1.copy(), tempDict2.copy()
```

Comment:
Line 281 suggest Domain-style dicts are nested, specifically `config` key. It might be worth seeing about using the `copy.deepcopy()` method from the `copy` module, to apply it down the dictionary layers.

Reply:
Right, it gets resolved earlier in the stream into a single tier dict. Got it.